File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptor.java
/*
 * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
 *
 * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Modifications copyright (C) 2017 Uber Technologies, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this material except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.temporal.common.interceptors;

import io.temporal.activity.ActivityOptions;
import io.temporal.activity.LocalActivityOptions;
import io.temporal.api.common.v1.WorkflowExecution;
import io.temporal.common.Experimental;
import io.temporal.common.SearchAttributeUpdate;
import io.temporal.workflow.*;
import io.temporal.workflow.Functions.Func;
import java.lang.reflect.Type;
import java.time.Duration;
import java.util.*;
import java.util.function.BiPredicate;
import java.util.function.Supplier;
import javax.annotation.Nullable;

/**
 * Can be used to intercept calls from to workflow code into the Temporal APIs.
 *
 * <p>The calls to the interceptor are executed in the context of a workflow and must follow the
 * same rules all the other workflow code follows.
 *
 * <p>Prefer extending {@link WorkflowOutboundCallsInterceptorBase} and overriding only the methods
 * you need instead of implementing this interface directly. {@link
 * WorkflowOutboundCallsInterceptorBase} provides correct default implementations to all the methods
 * of this interface.
 *
 * <p>An instance may be created in {@link
 * WorkflowInboundCallsInterceptor#init(WorkflowOutboundCallsInterceptor)} and set by passing it
 * into {@code init} method of the {@code next} {@link WorkflowInboundCallsInterceptor} The
 * implementation must forward all the calls to the outbound interceptor passed as a {@code
 * outboundCalls} parameter to the {@code init} call.
 *
 * @see WorkerInterceptor#interceptWorkflow for the definition of "next" {@link
 *     WorkflowInboundCallsInterceptor}.
 */
@Experimental
public interface WorkflowOutboundCallsInterceptor {

  final class ActivityInput<R> {
    private final String activityName;
    private final Class<R> resultClass;
    private final Type resultType;
    private final Object[] args;
    private final ActivityOptions options;
    private final Header header;

    public ActivityInput(
        String activityName,
        Class<R> resultClass,
        Type resultType,
        Object[] args,
        ActivityOptions options,
        Header header) {
      this.activityName = activityName;
      this.resultClass = resultClass;
      this.resultType = resultType;
      this.args = args;
      this.options = options;
      this.header = header;
    }

    public String getActivityName() {
      return activityName;
    }

    public Class<R> getResultClass() {
      return resultClass;
    }

    public Type getResultType() {
      return resultType;
    }

    public Object[] getArgs() {
      return args;
    }

    public ActivityOptions getOptions() {
      return options;
    }

    public Header getHeader() {
      return header;
    }
  }

  final class ActivityOutput<R> {
    private final String activityId;
    private final Promise<R> result;

    public ActivityOutput(String activityId, Promise<R> result) {
      this.activityId = activityId;
      this.result = result;
    }

    public String getActivityId() {
      return activityId;
    }

    public Promise<R> getResult() {
      return result;
    }
  }

  final class LocalActivityInput<R> {
    private final String activityName;
    private final Class<R> resultClass;
    private final Type resultType;
    private final Object[] args;
    private final LocalActivityOptions options;
    private final Header header;

    public LocalActivityInput(
        String activityName,
        Class<R> resultClass,
        Type resultType,
        Object[] args,
        LocalActivityOptions options,
        Header header) {
      this.activityName = activityName;
      this.resultClass = resultClass;
      this.resultType = resultType;
      this.args = args;
      this.options = options;
      this.header = header;
    }

    public String getActivityName() {
      return activityName;
    }

    public Class<R> getResultClass() {
      return resultClass;
    }

    public Type getResultType() {
      return resultType;
    }

    public Object[] getArgs() {
      return args;
    }

    public LocalActivityOptions getOptions() {
      return options;
    }

    public Header getHeader() {
      return header;
    }
  }

  final class LocalActivityOutput<R> {
    private final Promise<R> result;

    public LocalActivityOutput(Promise<R> result) {
      this.result = result;
    }

    public Promise<R> getResult() {
      return result;
    }
  }

  final class ChildWorkflowInput<R> {
    private final String workflowId;
    private final String workflowType;
    private final Class<R> resultClass;
    private final Type resultType;
    private final Object[] args;
    private final ChildWorkflowOptions options;
    private final Header header;

    public ChildWorkflowInput(
        String workflowId,
        String workflowType,
        Class<R> resultClass,
        Type resultType,
        Object[] args,
        ChildWorkflowOptions options,
        Header header) {
      this.workflowId = workflowId;
      this.workflowType = workflowType;
      this.resultClass = resultClass;
      this.resultType = resultType;
      this.args = args;
      this.options = options;
      this.header = header;
    }

    public String getWorkflowId() {
      return workflowId;
    }

    public String getWorkflowType() {
      return workflowType;
    }

    public Class<R> getResultClass() {
      return resultClass;
    }

    public Type getResultType() {
      return resultType;
    }

    public Object[] getArgs() {
      return args;
    }

    public ChildWorkflowOptions getOptions() {
      return options;
    }

    public Header getHeader() {
      return header;
    }
  }

  final class ChildWorkflowOutput<R> {

    private final Promise<R> result;
    private final Promise<WorkflowExecution> workflowExecution;

    public ChildWorkflowOutput(Promise<R> result, Promise<WorkflowExecution> workflowExecution) {
      this.result = result;
      this.workflowExecution = workflowExecution;
    }

    public Promise<R> getResult() {
      return result;
    }

    public Promise<WorkflowExecution> getWorkflowExecution() {
      return workflowExecution;
    }
  }

  @Experimental
  final class ExecuteNexusOperationInput<R> {
    private final String endpoint;
    private final String service;
    private final String operation;
    private final Class<R> resultClass;
    private final Type resultType;
    private final Object arg;
    private final NexusOperationOptions options;
    private final Map<String, String> headers;

    public ExecuteNexusOperationInput(
        String endpoint,
        String service,
        String operation,
        Class<R> resultClass,
        Type resultType,
        Object arg,
        NexusOperationOptions options,
        Map<String, String> headers) {
      this.endpoint = endpoint;
      this.service = service;
      this.operation = operation;
      this.resultClass = resultClass;
      this.resultType = resultType;
      this.arg = arg;
      this.options = options;
      this.headers = headers;
    }

    public String getService() {
      return service;
    }

    public String getOperation() {
      return operation;
    }

    public Object getArg() {
      return arg;
    }

    public Class<R> getResultClass() {
      return resultClass;
    }

    public Type getResultType() {
      return resultType;
    }

    public String getEndpoint() {
      return endpoint;
    }

    public NexusOperationOptions getOptions() {
      return options;
    }

    public Map<String, String> getHeaders() {
      return headers;
    }
  }

  @Experimental
  final class ExecuteNexusOperationOutput<R> {
    private final Promise<R> result;
    private final Promise<Optional<String>> operationExecution;

    public ExecuteNexusOperationOutput(
        Promise<R> result, Promise<Optional<String>> operationExecution) {
      this.result = result;
      this.operationExecution = operationExecution;
    }

    public Promise<R> getResult() {
      return result;
    }

    public Promise<Optional<String>> getOperationExecution() {
      return operationExecution;
    }
  }

  final class SignalExternalInput {
    private final WorkflowExecution execution;
    private final String signalName;
    private final Header header;
    private final Object[] args;

    public SignalExternalInput(
        WorkflowExecution execution, String signalName, Header header, Object[] args) {
      this.execution = execution;
      this.signalName = signalName;
      this.header = header;
      this.args = args;
    }

    public WorkflowExecution getExecution() {
      return execution;
    }

    public String getSignalName() {
      return signalName;
    }

    public Header getHeader() {
      return header;
    }

    public Object[] getArgs() {
      return args;
    }
  }

  final class SignalExternalOutput {
    private final Promise<Void> result;

    public SignalExternalOutput(Promise<Void> result) {
      this.result = result;
    }

    public Promise<Void> getResult() {
      return result;
    }
  }

  final class CancelWorkflowInput {
    private final WorkflowExecution execution;

    public CancelWorkflowInput(WorkflowExecution execution) {
      this.execution = execution;
    }

    public WorkflowExecution getExecution() {
      return execution;
    }
  }

  final class CancelWorkflowOutput {
    private final Promise<Void> result;

    public CancelWorkflowOutput(Promise<Void> result) {
      this.result = result;
    }

    public Promise<Void> getResult() {
      return result;
    }
  }

  final class ContinueAsNewInput {
    private final @Nullable String workflowType;
    private final @Nullable ContinueAsNewOptions options;
    private final Object[] args;
    private final Header header;

    public ContinueAsNewInput(
        @Nullable String workflowType,
        @Nullable ContinueAsNewOptions options,
        Object[] args,
        Header header) {
      this.workflowType = workflowType;
      this.options = options;
      this.args = args;
      this.header = header;
    }

    /**
     * @return workflowType for the continue-as-new workflow run. null if continue-as-new should
     *     inherit the type of the original workflow run.
     */
    public @Nullable String getWorkflowType() {
      return workflowType;
    }

    /**
     * @return options for the continue-as-new workflow run. Can be null, in that case the values
     *     will be taken from the original workflow run.
     */
    public @Nullable ContinueAsNewOptions getOptions() {
      return options;
    }

    public Object[] getArgs() {
      return args;
    }

    public Header getHeader() {
      return header;
    }
  }

  final class SignalRegistrationRequest {
    private final String signalType;
    private final HandlerUnfinishedPolicy unfinishedPolicy;
    private final Class<?>[] argTypes;
    private final Type[] genericArgTypes;
    private final Functions.Proc1<Object[]> callback;

    // Kept for backward compatibility
    public SignalRegistrationRequest(
        String signalType,
        Class<?>[] argTypes,
        Type[] genericArgTypes,
        Functions.Proc1<Object[]> callback) {
      this.signalType = signalType;
      this.unfinishedPolicy = HandlerUnfinishedPolicy.WARN_AND_ABANDON;
      this.argTypes = argTypes;
      this.genericArgTypes = genericArgTypes;
      this.callback = callback;
    }

    public SignalRegistrationRequest(
        String signalType,
        HandlerUnfinishedPolicy unfinishedPolicy,
        Class<?>[] argTypes,
        Type[] genericArgTypes,
        Functions.Proc1<Object[]> callback) {
      this.signalType = signalType;
      this.unfinishedPolicy = unfinishedPolicy;
      this.argTypes = argTypes;
      this.genericArgTypes = genericArgTypes;
      this.callback = callback;
    }

    public String getSignalType() {
      return signalType;
    }

    public HandlerUnfinishedPolicy getUnfinishedPolicy() {
      return unfinishedPolicy;
    }

    public Class<?>[] getArgTypes() {
      return argTypes;
    }

    public Type[] getGenericArgTypes() {
      return genericArgTypes;
    }

    public Functions.Proc1<Object[]> getCallback() {
      return callback;
    }
  }

  final class RegisterSignalHandlersInput {
    private final List<SignalRegistrationRequest> requests;

    public RegisterSignalHandlersInput(List<SignalRegistrationRequest> requests) {
      this.requests = requests;
    }

    public List<SignalRegistrationRequest> getRequests() {
      return requests;
    }
  }

  @Experimental
  final class UpdateRegistrationRequest {
    private final String updateName;
    private final HandlerUnfinishedPolicy unfinishedPolicy;
    private final Class<?>[] argTypes;
    private final Type[] genericArgTypes;
    private final Functions.Func1<Object[], Object> executeCallback;
    private final Functions.Proc1<Object[]> validateCallback;

    public UpdateRegistrationRequest(
        String updateName,
        HandlerUnfinishedPolicy unfinishedPolicy,
        Class<?>[] argTypes,
        Type[] genericArgTypes,
        Functions.Proc1<Object[]> validateCallback,
        Functions.Func1<Object[], Object> executeCallback) {
      this.updateName = updateName;
      this.unfinishedPolicy = unfinishedPolicy;
      this.argTypes = argTypes;
      this.genericArgTypes = genericArgTypes;
      this.validateCallback = validateCallback;
      this.executeCallback = executeCallback;
    }

    public String getUpdateName() {
      return updateName;
    }

    public HandlerUnfinishedPolicy getUnfinishedPolicy() {
      return unfinishedPolicy;
    }

    public Class<?>[] getArgTypes() {
      return argTypes;
    }

    public Type[] getGenericArgTypes() {
      return genericArgTypes;
    }

    public Functions.Proc1<Object[]> getValidateCallback() {
      return validateCallback;
    }

    public Functions.Func1<Object[], Object> getExecuteCallback() {
      return executeCallback;
    }
  }

  @Experimental
  final class RegisterUpdateHandlersInput {
    private final List<UpdateRegistrationRequest> requests;

    public RegisterUpdateHandlersInput(List<UpdateRegistrationRequest> requests) {
      this.requests = requests;
    }

    public List<UpdateRegistrationRequest> getRequests() {
      return requests;
    }
  }

  final class RegisterQueryInput {
    private final String queryType;
    private final Class<?>[] argTypes;
    private final Type[] genericArgTypes;
    private final Functions.Func1<Object[], Object> callback;

    public RegisterQueryInput(
        String queryType,
        Class<?>[] argTypes,
        Type[] genericArgTypes,
        Functions.Func1<Object[], Object> callback) {
      this.queryType = queryType;
      this.argTypes = argTypes;
      this.genericArgTypes = genericArgTypes;
      this.callback = callback;
    }

    public String getQueryType() {
      return queryType;
    }

    public Class<?>[] getArgTypes() {
      return argTypes;
    }

    public Type[] getGenericArgTypes() {
      return genericArgTypes;
    }

    public Functions.Func1<Object[], Object> getCallback() {
      return callback;
    }
  }

  final class RegisterDynamicQueryHandlerInput {
    private final DynamicQueryHandler handler;

    public RegisterDynamicQueryHandlerInput(DynamicQueryHandler handler) {
      this.handler = handler;
    }

    public DynamicQueryHandler getHandler() {
      return handler;
    }
  }

  final class RegisterDynamicSignalHandlerInput {
    private final DynamicSignalHandler handler;

    public RegisterDynamicSignalHandlerInput(DynamicSignalHandler handler) {
      this.handler = handler;
    }

    public DynamicSignalHandler getHandler() {
      return handler;
    }
  }

  @Experimental
  final class RegisterDynamicUpdateHandlerInput {
    private final DynamicUpdateHandler handler;

    public RegisterDynamicUpdateHandlerInput(DynamicUpdateHandler handler) {
      this.handler = handler;
    }

    public DynamicUpdateHandler getHandler() {
      return handler;
    }
  }

  <R> ActivityOutput<R> executeActivity(ActivityInput<R> input);

  <R> LocalActivityOutput<R> executeLocalActivity(LocalActivityInput<R> input);

  <R> ChildWorkflowOutput<R> executeChildWorkflow(ChildWorkflowInput<R> input);

  @Experimental
  <R> ExecuteNexusOperationOutput<R> executeNexusOperation(ExecuteNexusOperationInput<R> input);

  Random newRandom();

  SignalExternalOutput signalExternalWorkflow(SignalExternalInput input);

  CancelWorkflowOutput cancelWorkflow(CancelWorkflowInput input);

  void sleep(Duration duration);

  boolean await(Duration timeout, String reason, Supplier<Boolean> unblockCondition);

  void await(String reason, Supplier<Boolean> unblockCondition);

  Promise<Void> newTimer(Duration duration);

  <R> R sideEffect(Class<R> resultClass, Type resultType, Func<R> func);

  <R> R mutableSideEffect(
      String id, Class<R> resultClass, Type resultType, BiPredicate<R, R> updated, Func<R> func);

  int getVersion(String changeId, int minSupported, int maxSupported);

  void continueAsNew(ContinueAsNewInput input);

  void registerQuery(RegisterQueryInput input);

  void registerSignalHandlers(RegisterSignalHandlersInput input);

  @Experimental
  void registerUpdateHandlers(RegisterUpdateHandlersInput input);

  void registerDynamicSignalHandler(RegisterDynamicSignalHandlerInput handler);

  void registerDynamicQueryHandler(RegisterDynamicQueryHandlerInput input);

  @Experimental
  void registerDynamicUpdateHandler(RegisterDynamicUpdateHandlerInput input);

  UUID randomUUID();

  void upsertSearchAttributes(Map<String, ?> searchAttributes);

  void upsertTypedSearchAttributes(SearchAttributeUpdate<?>... searchAttributeUpdates);

  void upsertMemo(Map<String, Object> memo);

  /**
   * Intercepts creation of a workflow child thread.
   *
   * <p>Please note, that "workflow child thread" and "child workflow" are different and independent
   * concepts.
   *
   * @param runnable thread function to run
   * @param detached if this thread is detached from the parent {@link CancellationScope}
   * @param name name of the thread
   * @return created WorkflowThread
   */
  Object newChildThread(Runnable runnable, boolean detached, String name);

  long currentTimeMillis();
}

File: temporal-kotlin/src/main/kotlin/io/temporal/kotlin/interceptors/WorkflowOutboundCallsInterceptor.kt
/*
 * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
 *
 * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Modifications copyright (C) 2017 Uber Technologies, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this material except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.temporal.kotlin.interceptors

import io.temporal.activity.ActivityOptions
import io.temporal.activity.LocalActivityOptions
import io.temporal.api.common.v1.WorkflowExecution
import io.temporal.common.Experimental
import io.temporal.common.SearchAttributeUpdate
import io.temporal.common.interceptors.Header
import io.temporal.workflow.ChildWorkflowOptions
import io.temporal.workflow.ContinueAsNewOptions
import io.temporal.workflow.DynamicQueryHandler
import io.temporal.workflow.DynamicSignalHandler
import io.temporal.workflow.DynamicUpdateHandler
import io.temporal.workflow.Functions.Func
import io.temporal.workflow.Functions.Func1
import io.temporal.workflow.Functions.Proc1
import io.temporal.workflow.Promise
import java.lang.reflect.Type
import java.time.Duration
import java.util.*
import java.util.function.BiPredicate
import java.util.function.Supplier

/**
 * Can be used to intercept calls from to workflow code into the Temporal APIs.
 *
 *
 * The calls to the interceptor are executed in the context of a workflow and must follow the
 * same rules all the other workflow code follows.
 *
 *
 * Prefer extending [WorkflowOutboundCallsInterceptorBase] and overriding only the methods
 * you need instead of implementing this interface directly. [ ] provides correct default implementations to all the methods
 * of this interface.
 *
 *
 * An instance may be created in [ ][WorkflowInboundCallsInterceptor.init] and set by passing it
 * into `init` method of the `next` [WorkflowInboundCallsInterceptor] The
 * implementation must forward all the calls to the outbound interceptor passed as a `outboundCalls` parameter to the `init` call.
 *
 * @see KotlinWorkerInterceptor.interceptWorkflow for the definition of "next" {@link
 * *     WorkflowInboundCallsInterceptor}.
 */
@Experimental
interface WorkflowOutboundCallsInterceptor {
  class ActivityInput<R>(
    val activityName: String,
    val resultClass: Class<R>,
    val resultType: Type,
    val args: Array<out Any>,
    val options: ActivityOptions,
    val header: Header
  )

  class ActivityOutput<R>(val activityId: String, val result: R?)

  class LocalActivityInput<R>(
    val activityName: String,
    val resultClass: Class<R>,
    val resultType: Type,
    val args: Array<out Any>,
    val options: LocalActivityOptions,
    val header: Header
  )

  class LocalActivityOutput<R>(val result: R?)
  class ChildWorkflowInput<R>(
    val workflowId: String,
    val workflowType: String,
    val resultClass: Class<R>,
    val resultType: Type,
    val args: Array<Any>,
    val options: ChildWorkflowOptions,
    val header: Header
  )

  class ChildWorkflowOutput<R>(val result: Promise<R>, val workflowExecution: Promise<WorkflowExecution>)
  class SignalExternalInput(val execution: WorkflowExecution, val signalName: String, val args: Array<Any>)
  class SignalExternalOutput(val result: Promise<Void>)
  class CancelWorkflowInput(val execution: WorkflowExecution)
  class CancelWorkflowOutput(val result: Promise<Void>)
  class ContinueAsNewInput(
    /**
     * @return workflowType for the continue-as-new workflow run. null if continue-as-new should
     * inherit the type of the original workflow run.
     */
    val workflowType: String?,
    /**
     * @return options for the continue-as-new workflow run. Can be null, in that case the values
     * will be taken from the original workflow run.
     */
    val options: ContinueAsNewOptions?,
    val args: Array<Any>,
    val header: Header
  )

  class SignalRegistrationRequest(
    val signalType: String,
    val argTypes: Array<Class<*>>,
    val genericArgTypes: Array<Type>,
    val callback: Proc1<Array<Any>>
  )

  class RegisterSignalHandlersInput(val requests: List<SignalRegistrationRequest>)

  @Experimental
  class UpdateRegistrationRequest(
    val updateName: String,
    val argTypes: Array<Class<*>>,
    val genericArgTypes: Array<Type>,
    val validateCallback: Proc1<Array<Any>>,
    val executeCallback: Func1<Array<Any>, Any>
  )

  @Experimental
  class RegisterUpdateHandlersInput(val requests: List<UpdateRegistrationRequest>)
  class RegisterQueryInput(
    val queryType: String,
    val argTypes: Array<Class<*>>,
    val genericArgTypes: Array<Type>,
    val callback: Func1<Array<Any>, Any>
  )

  class RegisterDynamicQueryHandlerInput(val handler: DynamicQueryHandler)
  class RegisterDynamicSignalHandlerInput(val handler: DynamicSignalHandler)

  @Experimental
  class RegisterDynamicUpdateHandlerInput(val handler: DynamicUpdateHandler)

  suspend fun <R> executeActivity(input: ActivityInput<R>): ActivityOutput<R?>
  suspend fun <R> executeLocalActivity(input: LocalActivityInput<R>): LocalActivityOutput<R?>
  suspend fun <R> executeChildWorkflow(input: ChildWorkflowInput<R>): ChildWorkflowOutput<R?>
  fun newRandom(): Random
  suspend fun signalExternalWorkflow(input: SignalExternalInput): SignalExternalOutput
  fun cancelWorkflow(input: CancelWorkflowInput): CancelWorkflowOutput

  // TODO: Consider removing sleep and keep only built in delay
  suspend fun sleep(duration: Duration)
  suspend fun await(timeout: Duration, reason: String?, unblockCondition: Supplier<Boolean?>): Boolean
  suspend fun await(reason: String?, unblockCondition: Supplier<Boolean?>)
  fun <R> sideEffect(resultClass: Class<R>, resultType: Type, func: Func<R?>): R?
  fun <R> mutableSideEffect(
    id: String,
    resultClass: Class<R>,
    resultType: Type,
    updated: BiPredicate<R?, R?>,
    func: Func<R?>
  ): R?

  fun getVersion(changeId: String, minSupported: Int, maxSupported: Int): Int
  fun continueAsNew(input: ContinueAsNewInput)
  fun registerQuery(input: RegisterQueryInput)
  fun registerSignalHandlers(input: RegisterSignalHandlersInput)

  @Experimental
  fun registerUpdateHandlers(input: RegisterUpdateHandlersInput)
  fun registerDynamicSignalHandler(input: RegisterDynamicSignalHandlerInput)
  fun registerDynamicQueryHandler(input: RegisterDynamicQueryHandlerInput)

  @Experimental
  fun registerDynamicUpdateHandler(input: RegisterDynamicUpdateHandlerInput)
  fun randomUUID(): UUID
  fun upsertSearchAttributes(searchAttributes: Map<String?, *>)
  fun upsertTypedSearchAttributes(searchAttributeUpdates: List<SearchAttributeUpdate<*>>)

  fun currentTimeMillis(): Long
}
